src/fluree/db/reasoner/owl_datalog.cljc

```diff
@@ -552,7 +976,28 @@
   (reduce
    (fn [acc union-of-statement]
      (let [unions            (util/unwrap-list (get union-of-statement const/iri-owl:unionOf))
-           {:keys [classes has-value]} (group-by equiv-class-type unions)
+           ;; Process each union member to extract classes (including from nested unions)
+           expanded-classes  (reduce (fn [acc* union-member]
+                                       (cond
+                                         ;; Simple class
+                                         (and (util/get-id union-member)
+                                              (not (contains? union-member const/iri-owl:unionOf)))
+                                         (conj acc* (util/get-id union-member))
+
+                                         ;; Nested union - flatten it
+                                         (contains? union-member const/iri-owl:unionOf)
+                                         (let [nested-unions (util/unwrap-list (get union-member const/iri-owl:unionOf))]
+                                           (reduce (fn [acc** nested]
+                                                     (if-let [nested-id (util/get-id nested)]
+                                                       (conj acc** nested-id)
+                                                       acc**))
+                                                   acc*
+                                                   nested-unions))
+
+                                         :else acc*))
+                                     []
+                                     unions)
+           {:keys [has-value]} (group-by equiv-class-type unions)
            restrictions      (cond->> []
                                has-value (into (has-value-condition "?y" has-value)))
```

Comment:
I'm not sure why we need to call `into` here with an empty vector, we'll never add any data to the result of `has-value-condition` regardless of the value of `has-value`.